src/hooks.ts

```diff
@@ -37,20 +37,26 @@ export function useUpdate<T>(
   const { invalidate } = useContext(stateContext)
   const localRef = useRef()
   const ref = optionalRef ? optionalRef : localRef
+  const prevDependentsRef = useRef(dependents)
+
+  useEffect(() => {
+    prevDependentsRef.current = dependents
+  })
+
   useLayoutEffect(() => {
-    if (ref.current) {
+    if (ref.current && prevDependentsRef.current !== dependents) {
       callback(ref.current)
       invalidate()
     }
-  }, dependents)
+  }, [callback, dependents, invalidate, ref])
   return ref
 }
```

Comment:
One way to still use dependencies that are passed in is to follow the `usePrevious` model. Something to note is that hook dependencies use shallow comparison checks which doesn't often play nicely with arrays since mutating an array doesn't cause a new object ref.